src/engine/renderer/InternalImage.cpp

```diff
@@ -0,0 +1,183 @@
+      derived from this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+DISCLAIMED. IN NO EVENT SHALL DAEMON DEVELOPERS BE LIABLE FOR ANY
+DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+(INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+===========================================================================
+*/
+// InternalImage.cpp
+#include "tr_local.h"
+
+// Comments may include short quotes from other authors.
+
+/*
+===============
+R_GetImageCustomScalingStep
+
+Returns amount of downscaling step to be done on the given image
+to perform optional picmip or other downscaling operation.
+
+- the r_picMip cvar tells the renderer to downscale textures as many time as this cvar value;
+- the imageMaxDimension material keyword tells the renderer to downscale related textures
+  to that dimension if natively larger than that dimension;
+- the r_imageMaxDimension cvar tells the renderer to downscale all textures to that dimension
+  if natively larger than that dimension;
+- the r_ignoreMaterialMaxDimension cvar tells the renderer to ignore the imageMaxDimension material
+  keyword, r_imageMaxDimension still applies;
+- the imageMinDimension material keyword tells the renderer to not downscale related textures
+  below that dimension, has priority over r_picMip and r_imageMaxDimension;
+- the r_imageMinDimension cvar tells the renderer to not downscale textures below that dimension,
+  has priority over r_picMip and r_imageMaxDimension;
+- the r_ignoreMaterialMinDemension cvar tells the renderer to ignore the imageMinDimension material
+  keyword, r_imageMinDimension still applies;
+- the r_replaceMaterialMinDimensionIfPresentWithMaxDimension cvar tells the renderer to use the
+  maximum texture size for textures having a material keyword setting a minimum dimension;
+
+Examples of scenarii:
+
+- r_picMip set 3 with some materials having imageMinDimension set to 128:
+  attempt to reduce three times the images, but for materials having that keyword,
+  stop before the image dimension would be smaller than 128;
+- r_imageMaxDimension set to 64 with some materials having imageMinDimension set to 128:
+  reduce images having dimension greater than 64 to fit 64 dimension, but for materials
+  having that keyword, reduce image to fit 128 dimension;
+- r_imageMaxDimension set to 64 with some materials having imageMinDimension set to 32:
```

Comment:
This example seems wrong - I don't see why anything is reduced to 32.

Reply:
Hmm, I meant `imageMaxDimension set to 32`:

```
- r_imageMaxDimension set to 64 with some materials having imageMaxDimension set to 32:
  reduce images having dimension greater than 64 to fit 64 dimension, but for materials
  having that keyword, reduce image to fit 32 dimension;
  ```